sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp/receiver.py

```diff
@@ -108,37 +146,77 @@ def _outgoing_disposition(
         settled: Optional[bool],
         state: Optional[Union[Received, Accepted, Rejected, Released, Modified]],
         batchable: Optional[bool],
+        *,
+        on_disposition: Optional[Callable] = None,
     ):
         if delivery_tag not in self._received_delivery_tags:
             raise AMQPException(condition=ErrorCondition.IllegalState, description="Delivery tag not found.")
 
         disposition_frame = DispositionFrame(
             role=self.role, first=first, last=last, settled=settled, state=state, batchable=batchable
         )
+
         if self.network_trace:
             _LOGGER.debug("-> %r", DispositionFrame(*disposition_frame), extra=self.network_trace_params)
+
+        # Track dispositions for settling messages
+        if on_disposition:
+            delivery = PendingDisposition(
+                on_delivery_settled=on_disposition,
+                frame=disposition_frame,
+                settled=settled,
+                transfer_state=state,
+                start=time.time(),
+                sent=True,
+            )
+            self._pending_receipts.append(delivery)
+
         self._session._outgoing_disposition(disposition_frame)  # pylint: disable=protected-access
         self._received_delivery_tags.remove(delivery_tag)
 
+    def _incoming_disposition(self, frame):
+        # If delivery_id is not settled, return
+        # if not frame[3]:  # settled
+        #     return
+        # TODO: would we ever receive a not settled disposition?
+        range_end = (frame[2] or frame[1]) + 1  # first or last
+        settled_ids = list(range(frame[1], range_end))
+        unsettled = []
+        for delivery in self._pending_receipts:
+            if delivery.sent and delivery.frame[1] in settled_ids:
+                delivery.on_settled(LinkDeliverySettleReason.DISPOSITION_RECEIVED, frame[4])  # state
+                continue
+            unsettled.append(delivery)
+        self._pending_receipts = unsettled
```

Comment:
should the modifications for `self._pending_receipts` be under a lock, esp with Swathis latest changes ? For example when ALR runs

Reply:
my thought was keep alive was removed so I removed it ---- I havent tested this with alr need to do that I wonder if that helped there as well or if it needs to remain whoops